aiida/tools/groups/paths.py

```diff
@@ -0,0 +1,351 @@
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        """Compare equality of path and type string to another ``GroupPath`` object."""
+        if not isinstance(other, GroupPath):
+            return NotImplemented
+        return (self.path, self.type_string) == (other.path, other.type_string)
+
+    def __lt__(self, other):
+        # type: (Any) -> bool
+        """Compare less-than operator of path and type string to another ``GroupPath`` object."""
+        if not isinstance(other, GroupPath):
+            return NotImplemented
+        return (self.path, self.type_string) < (other.path, other.type_string)
+
+    @property
+    def path(self):
+        # type: () -> str
+        """Return the path string."""
+        return self._path_string
+
+    @property
+    def path_list(self):
+        # type: () -> List[str]
+        """Return a list of the path components."""
+        return self._path_list[:]
+
+    @property
+    def key(self):
+        # type: () -> str
+        """Return the final component of the the path."""
+        if self._path_list:
+            return self._path_list[-1]
+        return None
+
+    @property
+    def delimiter(self):
+        # type: () -> str
+        """Return the delimiter used to split path into components."""
+        return self._delimiter
+
+    @property
+    def type_string(self):
+        # type: () -> str
+        """Return the type_string used to query for and instantiate a ``Group`` with."""
+        return self._type_string
+
+    @property
+    def parent(self):
+        # type: () -> Optional[GroupPath]
+        """Return the parent path."""
+        if self.path_list:
+            return GroupPath(
+                self.delimiter.join(self.path_list[:-1]),
+                type_string=self.type_string,
+                warn_invalid_child=self._warn_invalid_child
+            )
+        return None
+
+    def __truediv__(self, path):
+        # type: (str) -> GroupPath
+        """Return a child ``GroupPath``, with a new path formed by appending ``path`` to the current path."""
+        if not isinstance(path, str):
+            raise TypeError('path is not a string: {}'.format(path))
+        path = self._validate_path(path)
+        child = GroupPath(
+            path=self.path + self.delimiter + path if self.path else path,
+            type_string=self.type_string,
+            warn_invalid_child=self._warn_invalid_child
+        )
+        return child
+
+    def __getitem__(self, path):
+        # type: (str) -> GroupPath
+        """Return a child ``GroupPath``, with a new path formed by appending ``path`` to the current path."""
+        return self.__truediv__(path)
+
+    @property
+    def group(self):
+        # type: () -> Optional[orm.Group]
+        """Return the concrete group associated with this path."""
+        try:
+            if self.type_string is not None:
+                return orm.Group.objects.get(label=self.path)
```

Comment:
I just noticed a potential problem with this implementation. Consider the following
```
Group(label='a', type_string='user').store()
Group(label='a', type_string='auto.import').store()
path = GroupPath('a', type_string=None)
path.group  # This will except because the filters do not uniquely determine a single group.
```
You catch the `NotExistent` case but not the `MultipleObjectsError` that will be thrown in my example. Did you think about this and made the decision to ignore it, or is this something that wasn't discussed during the lifetime of the PR? I am pointing this out here, because quite a bit of the rest of the functionality may depend on this.

To be honest, I am not quite sure what should happen here. Maybe we should make the `type_string` required. In this way, one can only use the `GroupPath` to iterate over a specific type of groups, but at least you know that the `label` will guarantee uniqueness then.

Reply:
No I wouldn't say it has been specifically addressed. I agree that we err on the side of disallowing a `None` `type_string` for now, and this can always be addressed in a later PR.

On a related note, how does your group sub-classing PR discriminate group classes; *via* the type string?

Reply:
see 5141b0d

Reply:
Yes, exactly, the `type_string` is what is used to load the correct sub class. I am just now actually working on updating the `QueryBuilder` such that querying with subclasses works properly and this is also based on the `type_string` which itself is determined by the entry point name. However, working on this made me realize that using `core` as the entry point name for the base `Group` class, is not ideal, since this will never match any groups whose entry point that does not start with `core.`, which is to say all groups from external plugins... Not sure what the best approach is here. Setting the entry point for `Group` base class to empty string is not really possible I think, so I will just have to make an exception for it in the query builder filtering logic.

Reply:
👍 
This is the use case, where I could envisage you might want to start using `GroupPath` with multiple `type_string`. But yeh, I don't yet have a concrete solution for how to deal with path clashes in that instance.

Reply:
so in 22ee75a, as discussed, I have changed the `group` property to the `get_group` method. This will allow for the later extension of replacing `type_string` to `type_cls`, at which point the association of a path to only 0 or 1 groups cannot be guaranteed.
Within this extension, all "iterative" methods would still work fine (like `children` and `walk`), but for methods where you directly access a group; `get_group`, delete_group`, `get_or_create_group`, you would then want to add arguments/logic to ensure only one group is selected.